pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -0,0 +1,549 @@
+	name string,
+	informerIndex krt.Index[cluster.ID, krt.Collection[T]],
+	opts krt.OptionsBuilder,
+) krt.TransformationSingle[Cluster, krt.Collection[config.ObjectWithCluster[T]]] {
+	return func(ctx krt.HandlerContext, c Cluster) *krt.Collection[config.ObjectWithCluster[T]] {
+		clients := informerIndex.Lookup(c.ID)
+		if len(clients) == 0 {
+			return nil
+		}
+		objWithCluster := krt.NewCollection(clients[0], func(ctx krt.HandlerContext, obj T) *config.ObjectWithCluster[T] {
+			return &config.ObjectWithCluster[T]{ClusterID: c.ID, Object: &obj}
+		}, opts.With(
+			krt.WithName(name+"WithCluster"),
+			krt.WithMetadata(krt.Metadata{
+				ClusterKRTMetadataKey: c.ID,
+			}),
+		)...)
+		return &objWithCluster
+	}
+}
+
+func informerIndexByCluster[T controllers.ComparableObject](
+	informerCollection krt.Collection[krt.Collection[T]],
+) krt.Index[cluster.ID, krt.Collection[T]] {
+	return krt.NewIndex[cluster.ID, krt.Collection[T]](informerCollection, func(col krt.Collection[T]) []cluster.ID {
+		val, ok := col.Metadata()[ClusterKRTMetadataKey]
+		if !ok {
+			log.Warnf("Cluster metadata not set on collection %v", col)
+			return nil
+		}
+		id, ok := val.(cluster.ID)
+		if !ok {
+			log.Warnf("Invalid cluster metadata set on collection %v: %v", col, val)
+			return nil
+		}
+		return []cluster.ID{id}
+	})
+}
+
+func nestedCollectionIndexByCluster[T any](
+	collection krt.Collection[krt.Collection[config.ObjectWithCluster[T]]],
+) krt.Index[cluster.ID, krt.Collection[config.ObjectWithCluster[T]]] {
+	return krt.NewIndex[cluster.ID, krt.Collection[config.ObjectWithCluster[T]]](collection, func(col krt.Collection[config.ObjectWithCluster[T]]) []cluster.ID {
+		val, ok := col.Metadata()[ClusterKRTMetadataKey]
+		if !ok {
+			log.Warnf("Cluster metadata not set on collection %v", col)
+			return nil
+		}
+		id, ok := val.(cluster.ID)
+		if !ok {
+			log.Warnf("Invalid cluster metadata set on collection %v: %v", col, val)
+			return nil
+		}
+		return []cluster.ID{id}
+	})
+}
+
+func mergeServiceInfosWithCluster(localClusterID cluster.ID) func(serviceInfos []config.ObjectWithCluster[model.ServiceInfo]) *config.ObjectWithCluster[model.ServiceInfo] {
+	return func(serviceInfos []config.ObjectWithCluster[model.ServiceInfo]) *config.ObjectWithCluster[model.ServiceInfo] {
+		if len(serviceInfos) == 0 {
+			return nil
+		}
+		if len(serviceInfos) == 1 {
+			return &serviceInfos[0]
+		}
+
+		var vips []*workloadapi.NetworkAddress
+		var sans []string
+		var base config.ObjectWithCluster[model.ServiceInfo]
+		for _, obj := range serviceInfos {
+			if obj.Object == nil {
+				continue
+			}
+
+			vips = append(vips, obj.Object.Service.GetAddresses()...)
+			sans = append(sans, obj.Object.Service.GetSubjectAltNames()...)
+			if obj.ClusterID == localClusterID {
+				// This is the base object we'll append to since it's local
+				base = obj
+			}
+		}
+
+		// TODO: Do we need to merge anything else?
```

Comment:
Goes with my other question, but we spent a lot of effort building out the waypoint collections and are now not using them..?

Reply:
This was one of the questions I had left after building all of this: there's nowhere to store things like a set of all the waypoint addresses for the service. Not unless we want to add new data structures to our data model. I'm sort of curious why we went away from, e.g. `model.Service` in Ambient since it has the sort of clustered context we need

Reply:
Fwiw in ambient initial implementation there was the proto + like 1 field I believe. so it's naturally expanded over time.

Reply:
Ok good to know we have some room to grow the data model

Reply:
This was a good point; if we're merging everything into 1 service, we probably don't need every waypoint reference?

Reply:
Ag actually, we need this in case there are remote waypoints but no local waypoint. Ztunnel will need to find the waypoint service in state based on a random waypoint's gateway address so we need to populate the merged waypoint service with all of them